Feature/kafka-batch-source

yaojiefeng opened the pull request:
Kafka batch source.

Todo:
Code is based on Camus, need refactoring.
Make pipeline name available to the plugin to not let user input the table name.
More test.

albertshau commented on src/main/java/co/cask/hydrator/KafkaBatchSource.java:
```diff
@@ -0,0 +1,579 @@
+    for (KafkaRequest request : finalRequests) {
+      KafkaKey key = offsetKeys.get(request);
+
+      if (key != null) {
```
this loop is really confusing to me, is this just copied from camus?

For key to not be null, it means the KafkaRequest shows up multiple times in finalRequests?

albertshau commented on src/main/java/co/cask/hydrator/KafkaKey.java:
```diff
@@ -0,0 +1,148 @@
+      }
+    }
+  }
+}
```
newline

albertshau commented on src/main/java/co/cask/hydrator/KafkaReader.java:
```diff
@@ -0,0 +1,158 @@
+
+
+/**
+ *
```
needs a javadoc

albertshau commented on src/main/java/co/cask/hydrator/KafkaReader.java:
```diff
@@ -0,0 +1,158 @@
+  /**
+   * Creates a fetch request.
+   */
+
```
remove empty line

albertshau commented on src/main/java/co/cask/hydrator/KafkaReader.java:
```diff
@@ -0,0 +1,158 @@
+      simpleConsumer.close();
+    }
+  }
+}
```
newline

albertshau commented on src/main/java/co/cask/hydrator/KafkaRecordReader.java:
```diff
@@ -0,0 +1,114 @@
+import java.io.IOException;
+
+/**
+ * Etl Record Reader.
```
update this javadoc, it's not an etl reader.

albertshau commented on src/main/java/co/cask/hydrator/KafkaRequest.java:
```diff
@@ -0,0 +1,149 @@
+    long endOffset = getLastOffset();
+    return (endOffset - offset) * avgMsgSize;
+  }
+}
```
newline

albertshau commented on src/main/java/co/cask/hydrator/KafkaSplit.java:
```diff
@@ -0,0 +1,95 @@
+    request = null;
+    return result;
+  }
+}
```
newline

albertshau commented on src/main/java/co/cask/hydrator/LeaderInfo.java:
```diff
@@ -0,0 +1,43 @@
+  public URI getUri() {
+    return uri;
+  }
+}
```
newline

albertshau commented on src/main/java/co/cask/hydrator/KafkaBatchSource.java:
```diff
@@ -0,0 +1,579 @@
+          offsetKeys.put(
+            request,
+            new KafkaKey(request.getTopic(), request.getLeaderId(), request.getPartition(), 0, request
+              .getOffset()));
```
`request.getOffset()` should be on the same line

albertshau commented on src/main/java/co/cask/hydrator/KafkaBatchSource.java:
```diff
@@ -0,0 +1,579 @@
+
+  private List<TopicMetadata> getKafkaMetadata(List<String> metaRequestTopics) {
+    String brokerString = config.getBrokers();
+    if (brokerString.isEmpty()) {
```
all validation should be done in a validate() method in the config. We can call it once, then the rest of the code can be sure that everything is valid.

albertshau commented on src/main/java/co/cask/hydrator/KafkaBatchSource.java:
```diff
@@ -0,0 +1,579 @@
+    if (brokerString.isEmpty()) {
+      throw new InvalidParameterException("kafka.brokers must contain at least one node");
+    }
+    List<String> brokers = Arrays.asList(brokerString.split("\\s*,\\s*"));
```
parsing the string into a list should be a method in the config.

albertshau commented on src/main/java/co/cask/hydrator/KafkaBatchSource.java:
```diff
@@ -0,0 +1,579 @@
+  private List<TopicMetadata> getKafkaMetadata(List<String> metaRequestTopics) {
+    String brokerString = config.getBrokers();
+    if (brokerString.isEmpty()) {
+      throw new InvalidParameterException("kafka.brokers must contain at least one node");
```
we've been using IllegalArgumentException for invalid config properties, since they are given by the user.

albertshau commented on src/main/java/co/cask/hydrator/KafkaBatchSource.java:
```diff
@@ -0,0 +1,579 @@
+    List<TopicMetadata> topicMetadataList = null;
+    while (i < brokers.size() && !fetchMetaDataSucceeded) {
+      SimpleConsumer consumer = createBrokerConsumer(brokers.get(i));
+      LOG.info(String.format("Fetching metadata from broker %s with client id %s for %d topic(s) %s", brokers.get(i),
```
seems more like a debug statement. An info statement would log the offsets for each topic partition.

albertshau commented on src/main/java/co/cask/hydrator/KafkaBatchSource.java:
```diff
@@ -0,0 +1,579 @@
+                             consumer.clientId(), metaRequestTopics.size(), metaRequestTopics));
+      try {
+        topicMetadataList = consumer.send(new TopicMetadataRequest(metaRequestTopics)).topicsMetadata();
+        fetchMetaDataSucceeded = true;
```
is it enough to just query one broker? Isn't the topic metadata spread out across all the brokers (I'm not sure, just asking)?

albertshau commented on src/main/java/co/cask/hydrator/KafkaBatchSource.java:
```diff
@@ -0,0 +1,579 @@
+  }
+
+  private SimpleConsumer createBrokerConsumer(String broker) {
+    if (!broker.matches(".+:\\d+")) {
```
validation should happen in a validate method in the config

albertshau commented on src/main/java/co/cask/hydrator/KafkaBatchSource.java:
```diff
@@ -0,0 +1,579 @@
+      throw new InvalidParameterException("The kakfa broker " + broker + " must follow address:port pattern");
+    }
+    String[] hostPort = broker.split(":");
+    return createSimpleConsumer(hostPort[0], Integer.valueOf(hostPort[1]));
```
if the port is not an integer we should throw an exception with a descriptive message.

albertshau commented on src/main/java/co/cask/hydrator/KafkaBatchSource.java:
```diff
@@ -0,0 +1,579 @@
+      if (latestOffsetResponse != null) {
+        earliestOffsetResponse = getLatestOffsetResponse(consumer, earliestOffsetInfo);
+      }
+      consumer.close();
```
should this be in a finally block?

albertshau commented on src/main/java/co/cask/hydrator/KafkaBatchSource.java:
```diff
@@ -0,0 +1,579 @@
+        if (earliestOffset == -1) {
+          earliestOffset = latestOffset;
+        }
+        LOG.info("earliest offset is {}", earliestOffset);
```
this doesn't seem like a useful info message for users. Should have one log message which includes the topic, partition, earliest offset and latest offset.

albertshau commented on src/main/java/co/cask/hydrator/KafkaBatchSource.java:
```diff
@@ -0,0 +1,579 @@
+      }
+      return offsetResponse;
+    } catch (Exception e) {
+      //
```
ignoring an exception is almost never the right thing to do.  When can this happen? Shouldn't it cause the job to fail?

albertshau commented on src/main/java/co/cask/hydrator/KafkaBatchSource.java:
```diff
@@ -0,0 +1,579 @@
+  }
+
+  private OffsetResponse getLatestOffsetResponse(SimpleConsumer consumer,
+                                                   Map<TopicAndPartition, PartitionOffsetRequestInfo> offsetInfo) {
```
fix indentation

albertshau commented on src/main/java/co/cask/hydrator/KafkaBatchSource.java:
```diff
@@ -0,0 +1,579 @@
+        consumer.getOffsetsBefore(new kafka.javaapi.OffsetRequest(offsetInfo, kafka.api.OffsetRequest.CurrentVersion(),
+                                                                  "client"));
+      if (offsetResponse.hasError()) {
+        throw new RuntimeException("offsetReponse has error.");
```
Imagine if you got this error message as a user. It doesn't give any useful information. You should check which TopicAndPartitions have an error, and what the error code is.

albertshau commented on src/main/java/co/cask/hydrator/KafkaInputFormat.java:
```diff
@@ -0,0 +1,61 @@
+    throws IOException, InterruptedException {
+    return new KafkaRecordReader();
+  }
+
```
almost all of the Kafka related logic that is currently in the plugin class should really be in this class. This is because this class should be re-usable by other code if needed. Most InputFormat classes have various static methods that jobs will call during setup.

albertshau commented on src/main/java/co/cask/hydrator/KafkaKey.java:
```diff
@@ -0,0 +1,148 @@
+  @Override
+  public int compareTo(KafkaKey o) {
+    if (partition != o.partition) {
+      return partition = o.partition;
```
was this copy pasted? This is assignment...

albertshau replied:
should be doing something like:

```
int comp = Integer.compare(partition, o.partition);
if (comp != 0) {
  return comp;
}
comp = Long.compare(offset, o.offset);
if (comp != 0) {
  return comp;
}
// etc
```

albertshau commented on src/main/java/co/cask/hydrator/KafkaReader.java:
```diff
@@ -0,0 +1,158 @@
+      ByteBuffer key = message.key();
+
+      if (payload == null) {
+        LOG.warn("Received message with null message.payload(): " + msgAndOffset);
```
include the topic and partition in the log message

albertshau commented on src/main/java/co/cask/hydrator/KafkaReader.java:
```diff
@@ -0,0 +1,158 @@
+      }
+      return true;
+    } catch (Exception e) {
+      LOG.warn("Exception generated during processing fetchResponse");
```
need more information in this log message. Everything we log should contain enough information for the user to do some sort of follow up investigation.

albertshau commented on src/main/java/co/cask/hydrator/KafkaRequest.java:
```diff
@@ -0,0 +1,149 @@
+    long[] endOffset = response.offsets(topic, partition);
+    consumer.close();
+    if (endOffset.length == 0) {
+      LOG.info("The exception is thrown because the latest offset retunred zero for topic : " + topic
```
'retunred' -> 'returned'. I'm confused by this error message, I think it's saying it didn't get any results while querying for the latest offset for the topic and partition.

Also, this is really weird behavior, it info logs but then guarantees an exception for array out of bounds will occur below?? Why not just throw an exception here?

albertshau replied:
also, the code to get the earliest offset doesn't do this check.

albertshau commented on src/test/java/co/cask/hydrator/PipelineTest.java:
```diff
@@ -126,50 +135,43 @@ public void testTextFileSourceAndMoveAction() throws Exception {
+    workflowManager.waitForRun(ProgramRunStatus.COMPLETED, 1, TimeUnit.MINUTES);
+    outputRecords = new HashSet<>();
+    outputRecords.addAll(MockSink.readOutput(outputManager));
+    Assert.assertEquals(6, outputRecords.size());
```
can't just check the number of results is as expected. You need to validate the actual records are as expected.

albertshau commented on src/test/java/co/cask/hydrator/PipelineTest.java:
```diff
@@ -126,50 +135,43 @@ public void testTextFileSourceAndMoveAction() throws Exception {
+    messages.put("f", "6,christopher,walken");
+    sendKafkaMessage("users", messages);
+    workflowManager.start();
+    TimeUnit.SECONDS.sleep(10);
```
don't sleep. You can just rely on waitForRun

albertshau commented on src/test/java/co/cask/hydrator/PipelineTest.java:
```diff
@@ -126,50 +135,43 @@ public void testTextFileSourceAndMoveAction() throws Exception {
+    Assert.assertEquals(3, outputRecords.size());
+    DataSetManager<KeyValueTable> kvTable = getDataset("testKafkaSource");
+    KeyValueTable table = kvTable.get();
+    byte [] offset = table.read(new TopicAndPartition("users", 0).toString());
```
byte[]

albertshau commented on src/test/java/co/cask/hydrator/PipelineTest.java:
```diff
@@ -126,50 +135,43 @@ public void testTextFileSourceAndMoveAction() throws Exception {
-                   .build());
-    Assert.assertEquals(expected, outputRecords);
-
+    Assert.assertEquals(3, outputRecords.size());
```
check the records are as expected, not just the number of records.

yaojiefeng commented:
@albertshau I have addressed most of your comments. For the logic regarding KafkaRequest and brokers, I need to investigate further as these are logics from Camus. The reason I put the logic which should be in KafkaInputFormat to KafkaBatchSource is that we need to read from the KVTable to get the start offset to run. I have now put all of them in KafkaInputFormat.

albertshau commented on docs/Kafka-batchsource.md:
```diff
@@ -3,9 +3,9 @@
-
+Kafka batch source. Emits the record from kafka. It will emit a record based on the schema and format 
+you use, or if no schema or format is specified, the message payload will be emitted. The source will 
+remember the offset it reads last time and continue from the offset for the next run.
```
remember the offset it read last run and continue from that offset for the next run

albertshau commented on src/main/java/co/cask/hydrator/KafkaRequest.java:
```diff
@@ -108,6 +109,10 @@ public long getEarliestOffset() {
         consumer.getOffsetsBefore(new OffsetRequest(offsetInfo, kafka.api.OffsetRequest.CurrentVersion(), "client"));
       long[] endOffset = response.offsets(topic, partition);
+      if (endOffset.length == 0) {
+        throw new RuntimeException("The exception is thrown because the latest offset returned zero for topic : " +
```
no need to say 'The exception is thrown because'. Just say 'Could not find earliest offset for topic x and partition y'

albertshau commented on src/main/java/co/cask/hydrator/KafkaRequest.java:
```diff
@@ -125,18 +130,16 @@ public long getLastOffset() {
     if (endOffset.length == 0) {
-      LOG.info("The exception is thrown because the latest offset retunred zero for topic : " + topic
-                 + " and partition " + partition);
+      throw new RuntimeException("The exception is thrown because the latest offset returned zero for topic : " + topic
```
'Could not find last offset for topic x and partition y'

albertshau commented on src/main/java/co/cask/hydrator/KafkaInputFormat.java:
```diff
@@ -109,7 +110,7 @@
       }
     } catch (Exception e) {
-      Throwables.propagate(e);
+      LOG.error("Error getting metadata for topic: {}", topic);
```
I did not mean to log and continue. I meant you don't need the try-catch. Failure to get topic metadata means the job will fail.

albertshau commented on src/main/java/co/cask/hydrator/KafkaInputFormat.java:
```diff
@@ -68,7 +69,7 @@
     Gson gson = new Gson();
-    List<KafkaRequest> finalRequests = gson.fromJson(context.getConfiguration().get(KafkaBatchSource.KAFKA_REQUEST),
+    List<KafkaRequest> finalRequests = gson.fromJson(context.getConfiguration().get(KAFKA_REQUEST),
                                                      LIST_TYPE);
```
nit: can this be one line now?

albertshau commented:
have we tested what happens if the wrong broker information is given so that it can't get the required info? Does it fail with a useful error message?

yaojiefeng commented:
@albertshau Addressed your comments, also added a check on whether we can get the required info for the topic. If we cannot get anything from the broker, an exception will be thrown with the topic and broker and fail the pipeline, pls take another look, thank you!
